amethyst_network/src/systems/network_client.rs

```diff
@@ -0,0 +1,119 @@
+//! The network client System
+
+use specs::{System,FetchMut};
+use std::net::UdpSocket;
+use std::net::IpAddr;
+use std::net::SocketAddr;
+use std::io::{Error,ErrorKind};
+use std::str;
+use std::str::FromStr;
+use std::clone::Clone;
+
+use shrev::*;
+
+use resources::*;
+use systems::NetworkBase;
+use std::marker::PhantomData;
+
+use serde::Serialize;
+use serde::de::DeserializeOwned;
+
+/// The System managing the client's network state and connections.
+/// The T generic parameter corresponds to the network event enum type.
+pub struct NetClientSystem<T> where T:Send+Sync{
+    /// The network socket
+    pub socket:UdpSocket,
+    /// The server that we are (possibly) connected or connecting to.
+    pub connection:Option<NetConnection>,//Will handle a single connection for now, as it is simple to manage and corresponds to most use cases.
+    net_event_types:PhantomData<T>,
+}
+
+//TODO: add Unchecked Event type list. Those events will be let pass the client connected filter (Example: NetEvent::Connect).
+//TODO: add different Filters that can be added on demand, to filter the event before they reach other systems.
+impl<T> NetClientSystem<T> where T:Send+Sync+Serialize+Clone+DeserializeOwned+BaseNetEvent<T>+'static{
+    /// Creates a NetClientSystem and binds the Socket on the ip and port added in parameters.
+    pub fn new(ip:&str,port:u16)->Result<NetClientSystem<T>,Error>{
+        let socket = UdpSocket::bind(SocketAddr::new(IpAddr::from_str(ip).expect("Unreadable input IP"),port))?;
+        socket.set_nonblocking(true)?;
+        Ok(
+            NetClientSystem{
+                socket,
+                connection:None,
+                net_event_types:PhantomData,
+            }
+        )
+    }
+    /// Connects to a remote server
+    pub fn connect(&mut self,target:SocketAddr){
+        println!("Sending connection request to remote {:?}",target);
+        let conn = NetConnection{
+            target,
+            state:ConnectionState::Connecting,
+        };
+        self.send_event(&T::base_to_custom(NetEvent::Connect),&conn,&self.socket);
+        self.connection = Some(
+            conn,
+        );
+    }
+}
+
+impl<T> NetworkBase<T> for NetClientSystem<T> where T:Send+Sync+Serialize+Clone+DeserializeOwned+BaseNetEvent<T>+'static{}
+
+impl<'a, T> System<'a> for NetClientSystem<T> where T:Send+Sync+Serialize+Clone+DeserializeOwned+BaseNetEvent<T>+'static{
+    type SystemData = (
+        FetchMut<'a, EventChannel<NetOwnedEvent<T>>>,
+    );
+    //omg unreadable plz enjoy code owo
+    fn run(&mut self, (mut events,): Self::SystemData) {
+        let mut buf = [0; 2048];
+        loop {
+            match self.socket.recv_from(&mut buf) {
+                //Data received
+                Ok((amt, src)) => {
+                    //Are we connected to anything?
+                    if self.connection.is_some(){
```

Comment:
`if let` is a cool construct that can help you out here by binding the inner value of the `Some`.

Reply:
because of borrowing rule I can't store connection in a local variable because I need to:
Read from self.connection
Replace self.connection
in that order and that won't work if self.connection is borrowed.

Reply:
Something like [this](https://play.rust-lang.org/?gist=364823074157ebeba22c7436f3b8cd13&version=stable) may or may not end up being cleaner.  Using the bool like that is kind of ugly but with how much unwrapping there is here it might actually be better.

Reply:
fixed